File: src/Structures/hs/Queue.hs
{-# LANGUAGE RecordWildCards #-}
module Queue where

import qualified Stack as Stack

-- | *Queue
data Queue a = Queue { queueFront :: Stack.Stack a
                     , queueBack :: Stack.Stack a
                     }
-- .end

-- | *empty
empty :: Queue a
empty = Queue { queueFront = Stack.empty
              , queueBack = Stack.empty
              }
-- .end

-- | *rotate
rotate :: Queue a -> Queue a
rotate Queue{..} =
  Queue { queueFront = Stack.reverse queueBack
        , queueBack = Stack.reverse queueFront
        }
-- .end

-- | *top
top :: Queue a -> Maybe a
top Queue{..} =
  case (Stack.head queueFront, Stack.head $ Stack.reverse queueBack) of
    (Nothing, Nothing) -> Nothing
    (Just x, _) -> Just x
    (Nothing, Just y) -> Just y
-- .end

-- | *back
back :: Queue a -> Maybe a
back Queue{..} =
  case (Stack.head queueBack, Stack.head $ Stack.reverse queueFront) of
    (Nothing, Nothing) -> Nothing
    (Just y, _) -> Just y
    (Nothing, Just x) -> Just x
-- .end

-- | *pop
pop :: Queue a -> Maybe (Queue a)
pop Queue{..} =
  case (Stack.tail queueFront, Stack.tail $ Stack.reverse queueBack) of
    (Nothing, Nothing) -> Nothing
    (Just xs, _) -> Just $ Queue {queueFront = xs, queueBack = queueBack}
    (Nothing, Just ys) -> Just $ Queue {queueFront = ys, queueBack = Stack.empty}
-- .end

-- | *eject
eject :: Queue a -> Maybe (Queue a)
eject Queue{..} =
  case (Stack.tail queueBack, Stack.tail $ Stack.reverse queueFront) of
    (Nothing, Nothing) -> Nothing
    (Just ys, _) -> Just $ Queue {queueBack = ys, queueFront = queueFront}
    (Nothing, Just xs) -> Just $ Queue {queueBack = xs, queueFront = Stack.empty}
-- .end

-- | *push
push :: a -> Queue a -> Queue a
push val Queue{..} =
  Queue { queueFront = Stack.cons val queueFront
        , queueBack = queueBack
        }
-- .end

-- | *inject
inject :: a -> Queue a -> Queue a
inject val Queue{..} =
  Queue { queueBack = Stack.cons val queueBack
        , queueFront = queueFront
        }
-- .end

File: src/Structures/scala/Queue.scala
package okasaki;

// | *Queue empty
object Queue
// .end
{
  // | *empty
  def empty = new Queue(Nil, Nil)
  // .end
}

// | *Queue isEmpty rotate top back pop push inject eject
sealed class Queue[+A](front: Stack[A], back: Stack[A])
// .end
{
  // | *isEmpty
  def isEmpty: Boolean =
    front.isEmpty && back.isEmpty
  // .end

  // | *rotate
  def rotate: Queue[A] =
    new Queue(back.reverse, front.reverse)
  // .end

  // | *top
  def top: A = front match {
    case Cons(h, _) => h
    case Nil => back.reverse match {
      case Cons(h, _) => h
      case Nil => throw new NoSuchElementException("Empty top")
    }
  }
  // .end

  // | *back
  def bottom: A = back match {
    case Cons(h, _) => h
    case Nil => front.reverse match {
      case Cons(h, _) => h
      case Nil => throw new NoSuchElementException("Empty back")
    }
  }
  // .end

  // | *pop
  def pop: Queue[A] = front match {
    case Cons(_, t) => new Queue(t, back)
    case Nil => back.reverse match {
      case Cons(_, t) => new Queue(t, Nil)
      case Nil => throw new NoSuchElementException("Empty pop")
    }
  }
  // .end

  // | *eject
  def eject: Queue[A] = back match {
    case Cons(_, t) => new Queue(front, t)
    case Nil => front.reverse match {
      case Cons(_, t) => new Queue(Nil, t)
      case Nil => throw new NoSuchElementException("Empty eject")
    }
  }
  // .end

  // | *push
  def push[B >: A](elem: B): Queue[B] =
    new Queue(front.cons(elem), back)
  // .end

  // | *inject
  def inject[B >: A](elem: B): Queue[B] =
    new Queue(front, back.cons(elem))
  // .end

  def shift[B >: A](elem: B): Queue[B] = inject(elem)
  def unshift: Queue[A] = pop

  override def toString = front.toString ++ " // " ++ back.toString
}
